src/graph/immutable_graph.cc

```diff
@@ -0,0 +1,975 @@
+  dgl_id_t* dst_data = static_cast<dgl_id_t*>(dst->data);
+  dgl_id_t* eid_data = static_cast<dgl_id_t*>(eid->data);
+  for (int64_t i = 0; i < len; ++i) {
+    src_data[i] = this->indices[off + i];
+    eid_data[i] = this->edge_ids[off + i];
+  }
+  std::fill(dst_data, dst_data + len, vid);
+  return ImmutableGraph::EdgeArray{src, dst, eid};
+}
+
+ImmutableGraph::EdgeArray ImmutableGraph::CSR::GetEdges(IdArray vids) const {
+  CHECK(IsValidIdArray(vids)) << "Invalid vertex id array.";
+  const auto len = vids->shape[0];
+  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
+  int64_t rstlen = 0;
+  for (int64_t i = 0; i < len; ++i) {
+    dgl_id_t vid = vid_data[i];
+    CHECK(HasVertex(vid)) << "Invalid vertex: " << vid;
+    rstlen += this->GetDegree(vid);
+  }
+  IdArray src = IdArray::Empty({rstlen}, vids->dtype, vids->ctx);
+  IdArray dst = IdArray::Empty({rstlen}, vids->dtype, vids->ctx);
+  IdArray eid = IdArray::Empty({rstlen}, vids->dtype, vids->ctx);
+  dgl_id_t* src_ptr = static_cast<dgl_id_t*>(src->data);
+  dgl_id_t* dst_ptr = static_cast<dgl_id_t*>(dst->data);
+  dgl_id_t* eid_ptr = static_cast<dgl_id_t*>(eid->data);
+  for (int64_t i = 0; i < len; ++i) {
+    dgl_id_t vid = vid_data[i];
+    int64_t off = this->indptr[vid];
+    const int64_t len = this->GetDegree(vid);
+    const auto *pred = &this->indices[off];
+    const auto *eids = &this->edge_ids[off];
+    for (int64_t j = 0; j < len; ++j) {
+      *(src_ptr++) = pred[j];
+      *(dst_ptr++) = vid;
+      *(eid_ptr++) = eids[j];
+    }
+  }
+  return ImmutableGraph::EdgeArray{src, dst, eid};
+}
+
+DegreeArray ImmutableGraph::CSR::GetDegrees(IdArray vids) const {
+  CHECK(IsValidIdArray(vids)) << "Invalid vertex id array.";
+  const auto len = vids->shape[0];
+  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
+  DegreeArray rst = DegreeArray::Empty({len}, vids->dtype, vids->ctx);
+  dgl_id_t* rst_data = static_cast<dgl_id_t*>(rst->data);
+  for (int64_t i = 0; i < len; ++i) {
+    const auto vid = vid_data[i];
+    CHECK(HasVertex(vid)) << "Invalid vertex: " << vid;
+    rst_data[i] = this->GetDegree(vid);
+  }
+  return rst;
+}
+
+class Bitmap {
+  const size_t size = 1024 * 1024 * 4;
+  const size_t mask = size - 1;
+  std::vector<bool> map;
+
+  size_t hash(dgl_id_t id) const {
+    return id & mask;
+  }
+ public:
+  Bitmap(const dgl_id_t *vid_data, int64_t len): map(size) {
+    for (int64_t i = 0; i < len; ++i) {
+      map[hash(vid_data[i])] = 1;
+    }
+  }
+
+  bool test(dgl_id_t id) const {
+    return map[hash(id)];
+  }
+};
+
+/*
+ * This uses a hashtable to check if a node is in the given node list.
+ */
+class HashTableChecker {
+  std::unordered_map<dgl_id_t, dgl_id_t> oldv2newv;
+  // This bitmap is used as a bloom filter to remove some lookups.
+  // Hashtable is very slow. Using bloom filter can significantly speed up lookups.
+  Bitmap map;
+
+  void Collect(const dgl_id_t old_id, const dgl_id_t old_eid,
```

Comment:
Could you give some brief docstring about what is this Collect and also the CollectOnRow is doing? Explain what's the meaning of the args. The docstring does not need to be as complete as the one in the interface, but it will help developers a lot later.